posts/airflow-2-scheduler.md

```diff
@@ -0,0 +1,204 @@
+
+Similarly, users can always reduce the number of schedulers to minimize resource usage as load lessens. Since each Scheduler is “fully active” and identical, there is no downside to scaling down based on the load.
+
+## Service Composition
+
+One of the distributed systems principles followed in the Airflow 2.0 Scheduler is that of _Service Composition_ to enable external tooling to manage the number of Scheduler instances to be run. This was a conscious choice leading to the following architectural decisions:
+
+* Having each Scheduler be complete and _“fully active”_, rather than have a _“leader”_ and other _“followers”_, with differing sets of capabilities. 
+* Not having internal monitoring of Scheduler instances by other Scheduler instances. This is generally, though not always, implemented by a _“leader election”_ model related to the above.
+* Expecting distributed system tooling to be _“layered”_ on top of Airflow to monitor elements such as task load and task latency, and to easily scale up / down Airflow Schedulers and Workers as needed.
+* Using the metadata database as the shared queue and synchronization mechanism. This has the added benefit of not requiring communication between Scheduler instances and therefore eliminates the need for additional system and network configuration.
+
+## Scalability Benchmarks
+
+We have been using _task throughput_ as the key metric for measuring Airflow scalability and to identify bottlenecks. _Task throughput_ is measured in “tasks per minute”. This represents the number of tasks that can be scheduled, queued, executed, and monitored by Airflow every minute.
+
+To sufficiently test this, without skewing numbers based on the actual task “work” time, we have chosen to benchmark using a simple PythonOperator task with a trivial execution time. The benchmarking configuration was: Celery Workers, PostgreSQL DB, 1 Web Server.
+
+Results for _task throughput_ (metric explained above) using Airflow 2.0 beta builds, run with 5,000 DAGs, each with 10 parallel tasks on a single Airflow deployment. The benchmark was performed on Google Cloud and each Scheduler was run on a [n1-standard-1 machine type](https://cloud.google.com/compute/docs/machine-types#n1_machine_types).
+
+| Number of Schedulers  | Number of Workers  | Task Throughput (average)  | Task Throughput (low) | Task Throughput (high) |
+|-----------------------|--------------------|----------------------------|-----------------------|------------------------|
+| 1                     | 12                 | 285                        | 248                   | 323                    |
+| 2                     | 12                 | 541                        | 492                   | 578                    |
+| 3                     | 12                 | 698                        | 632.5                 | 774                    |
+
+![Airflow Scheduler Task Throughput](..assets/airflow-2-scheduler/scheduler-task-throughput.png)
```

Comment:
the image url needs to be ../assets/airflow-2-scheduler/scheduler-task-throughput.png instead of ..assets/airflow-2-scheduler/scheduler-task-throughput.png